### ISSUE TITLE
loss increase when finetuning on shakespeare datasets with gpt-xl and mps device

### ISSUE THREAD

**Issue by davidlight2018** [karpathy/nanoGPT#367]
Hello, I've been attempting to fine-tune a GPT-XL model on a Shakespeare dataset using an MPS device. However, I've noticed that both the training loss and evaluation loss are significantly increasing. Additionally, the text generated by the model appears to be gibberish. Is this expected behavior? If not, what steps can I take to resolve these issues?

- samples paragraph
```
ptin,
 R- 5 long f 1 of 1 and- db.in very re ( like so for all or non under. k with to if that, up 3 ( no T we G- N, per some's by: would, C by in with,, such!.e... and would the's people with— T only for who but ( R like such
 any be. the I on all-—.. you '/ , with with at also out with for,.../s not the my of p and- Al or which down without every one had", of where. T or without three or  down R and (, up T being… M X [ righta & f St"
 at in a   any he. under how in is. under,- both. had is more two one ( e P F and L P high and off G new ts ( my an all T E " would with e W, of to -- from V here this where of and new my by Al a free/ a two is
 E every E his only 4 a not not when P how- L H... (., - be per) In almost D two 4 the a you and - 1 a
 a' — go after to in ' 3 or e 5 E and for de, also really my – W with on Al froms, into." –, — every . is an in both c un were which in or, " In " A, for- and H with: and also, its a… to St ( b... which every back in down I which either ( of have of " " free ''s an almost oring three on almost such over. they either to ... but either N said or In un up in/ al can "), " when and under here with by not the have use made so:/( every we back are on' were R, is . an because? use really ' no within in Nt G this can what were more would new — down my no will! by!. G R P and as such, The. of on about " and them in and a still-'s ( — they- on T off at out not after- as before free in back which 10 in un at I --'s is. of more ( do long ( under, and with al
. H to any about/ more non use " to New St one –
```
- loss log
```
step 0: train loss 3.9020, val loss 3.7291
iter 0: loss 3.6874, time 159524.25ms, mfu -100.00%
iter 1: loss 3.4146, time 79925.16ms, mfu -100.00%
iter 2: loss 4.8849, time 75545.46ms, mfu -100.00%
iter 3: loss 6.2328, time 75245.65ms, mfu -100.00%
iter 4: loss 7.7153, time 75968.43ms, mfu -100.00%
step 5: train loss 8.1077, val loss 7.9050
saving checkpoint to out-shakespeare
iter 5: loss 8.0483, time 174667.91ms, mfu 0.62%
iter 6: loss 8.3349, time 77925.59ms, mfu 0.69%
iter 7: loss 8.4360, time 75701.57ms, mfu 0.77%
iter 8: loss 8.6004, time 73011.09ms, mfu 0.84%
iter 9: loss 8.3439, time 74432.18ms, mfu 0.90%
step 10: train loss 8.3741, val loss 8.3617
saving checkpoint to out-shakespeare
iter 10: loss 8.4231, time 167228.05ms, mfu 0.87%
iter 11: loss 8.2665, time 72179.73ms, mfu 0.94%
iter 12: loss 8.4128, time 72144.53ms, mfu 0.99%
iter 13: loss 8.2871, time 70993.01ms, mfu 1.05%
iter 14: loss 8.3986, time 70279.50ms, mfu 1.09%
step 15: train loss 8.5151, val loss 8.5027
saving checkpoint to out-shakespeare
iter 15: loss 8.6310, time 166114.14ms, mfu 1.05%
iter 16: loss 8.6411, time 73397.54ms, mfu 1.09%
iter 17: loss 8.4767, time 73494.14ms, mfu 1.13%
iter 18: loss 8.6144, time 72748.71ms, mfu 1.17%
iter 19: loss 8.6268, time 72220.93ms, mfu 1.20%
step 20: train loss 8.6489, val loss 8.6528
saving checkpoint to out-shakespeare
iter 20: loss 8.6337, time 169740.58ms, mfu 1.14%
```

- config file
```
import time

out_dir = 'out-shakespeare'
eval_interval = 5
eval_iters = 40
wandb_log = True # feel free to turn on
wandb_project = 'shakespeare'
wandb_run_name = 'ft-' + str(time.time())

dataset = 'shakespeare'
init_from = 'gpt2-xl' # this is the largest GPT-2 model

# only save checkpoints if the validation loss improves
always_save_checkpoint = True

# the number of examples per iter:
# 1 batch_size * 32 grad_accum * 1024 tokens = 32,768 tokens/iter
# shakespeare has 301,966 tokens, so 1 epoch ~= 9.2 iters
batch_size = 1
gradient_accumulation_steps = 32
max_iters = 20

# finetune at constant LR
learning_rate = 3e-5
decay_lr = False

device = 'mps'
compile = True
```

**Comment by VatsaDev** [comment]
While I personally Don't own an MPS, and the highest I've trained till is gpt2-medium, but I have made `1.12` loss, I'd like to ask, Have you tried Increasing the learning rate, upping the batch size, or trying a bigger dataset?

For this repo, GPT2-XL is the biggest it gets, and Shakespeare is a very tiny Dataset
If none of this works, might just be that mps support isn't as great as the cuda support, could be a bug

**Comment by davidlight2018** [comment]
Thanks for the reply. After experimenting with CUDA and the gpt2-large model, I've managed to achieve more reliable results. Unfortunately, I can't experiment with the gpt2-XL model due to GPU memory limitations.
